cli/azd/pkg/auth/manager.go

```diff
@@ -0,0 +1,413 @@
+const authDirectoryFileMode = 0700
+
+// Manager manages the authentication system of azd. It allows a user to log in, either as a user principal or service
+// principal. Manager stores information so that the user can stay logged in across invocations of the CLI. When logged in
+// as a user (either interactively or via a device code flow), we provide a durable cache to MSAL which is used to cache
+// information to allow silent logins across process runs. This cache is stored inside the user's home directory, ACL'd such
+// that it can only be read by the current user.  In addition, on Windows, this cache is encrypted, using CryptProtectData.
+// The home account id of the signed in user is stored as a property under [cCurrentUserKey]. This behavior matches the
+// AZ CLI.
+//
+// When logged in as a service principal, the same cache strategy that backed the MSAL cache is used to store the private
+// key or secret and the public components (the client ID and tenant ID) are stored under  [cCurrentUserKey].
+//
+// Logging out removes this cached authentication data.
+//
+// You can configure azd to ignore its native credential system and instead delegate to AZ CLI (useful for cases where azd
+// does not yet support your preferred method of authentication by setting [cUseLegacyAzCliAuthKey] in config to true.
+type Manager struct {
+	publicClient    *public.Client
+	configManager   config.Manager
+	credentialCache exportReplaceWithErrors
+}
+
+func NewManager(configManager config.Manager) (*Manager, error) {
+	cfgRoot, err := config.GetUserConfigDir()
+	if err != nil {
+		return nil, fmt.Errorf("getting config dir: %w", err)
+	}
+
+	authRoot := filepath.Join(cfgRoot, "auth")
+	if err := os.MkdirAll(authRoot, authDirectoryFileMode); err != nil {
+		return nil, fmt.Errorf("creating auth root: %w", err)
+	}
+
+	cacheRoot := filepath.Join(authRoot, "msal")
+	if err := os.MkdirAll(cacheRoot, authDirectoryFileMode); err != nil {
+		return nil, fmt.Errorf("creating msal cache root: %w", err)
+	}
+
+	publicClientApp, err := public.New(cAZD_CLIENT_ID, public.WithCache(newCache(cacheRoot)))
+	if err != nil {
+		return nil, fmt.Errorf("creating msal client: %w", err)
+	}
+
+	return &Manager{
+		publicClient:    &publicClientApp,
+		configManager:   configManager,
+		credentialCache: newCredentialCache(authRoot),
+	}, nil
+}
+
+var ErrNoCurrentUser = errors.New("not logged in, run `azd login` to login")
+
+// EnsureLoggedInCredential uses the credential's GetToken method to ensure an access token can be fetched. If this fails,
+// nil, ErrNoCurrentUser is returned. On success, the token we fetched is returned.
+func EnsureLoggedInCredential(ctx context.Context, credential azcore.TokenCredential) (azcore.AccessToken, error) {
+	tok, err := credential.GetToken(ctx, policy.TokenRequestOptions{
+		Scopes: cLoginScopes,
+	})
+	if err != nil {
+		return azcore.AccessToken{}, ErrNoCurrentUser
+	}
+
+	return tok, nil
+}
```

Comment:
Consider using a pointer for the `azcore.AccessToken` so we can return a `nil` value in the error scenario.

Reply:
Changed - I actually thought about this initially but kept the non pointer return because it matched what the TokenCredential interface did for GetToken.  But I'm always happy to make the return value even harder to use in the bad case :-).